dsp/src/trig.rs

```diff
@@ -1,8 +1,85 @@
-use super::Complex;
+use super::{abs, shift_round, Complex};
 use core::f64::consts::PI;
 
 include!(concat!(env!("OUT_DIR"), "/cossin_table.rs"));
 
+/// 2-argument arctangent function.
+///
+/// This implementation uses all integer arithmetic for fast
+/// computation. It is designed to have high accuracy near the axes
+/// and lower away from the axes. It is additionally designed so that
+/// the error changes slowly with respect to the angle.
+///
+/// # Arguments
+///
+/// * `y` - Y-axis component.
+/// * `x` - X-axis component.
+///
+/// # Returns
+///
+/// The angle between the x-axis and the ray to the point (x,y). The
+/// result range is from i32::MIN to i32::MAX, where i32::MIN
+/// corresponds to an angle of -pi and i32::MAX corresponds to an
+/// angle of +pi.
```

Comment:
Is that really correct? `i32::MAX != -i32::MIN`.

Reply:
I think it's actually `i32::MIN == -pi` and at the same time also `i32::MIN == pi`. Then `i32::MAX == pi*(1 - 1/(1 << 31))`

Reply:
Oh, right that makes sense. It's also worth noting that `i32::MIN` is not actually reachable. I can get `i32::MAX` with `atan2(0, i32::MIN)`. The closest results are 2147461318 and -2147461318 with `atan2(1 << 16, i32::MIN))` and `atan2(-1 << 16, i32::MIN))`.